src/storage/data_table.cpp

```diff
@@ -122,23 +131,79 @@ bool DataTable::Update(transaction::TransactionContext *const txn, const TupleSl
+    insertion_head_ = --blocks_.end();
+  }
+}
+
 TupleSlot DataTable::Insert(transaction::TransactionContext *const txn, const ProjectedRow &redo) {
   TERRIER_ASSERT(redo.NumColumns() == accessor_.GetBlockLayout().NumColumns() - NUM_RESERVED_COLUMNS,
                  "The input buffer never changes the version pointer column, so it should have  exactly 1 fewer "
                  "attribute than the DataTable's layout.");
 
-  // Attempt to allocate a new tuple from the block we are working on right now.
-  // If that block is full, try to request a new block. Because other concurrent
-  // inserts could have already created a new block, we need to use compare and swap
-  // to change the insertion head. We do not expect this loop to be executed more than
-  // twice, but there is technically a possibility for blocks with only a few slots.
+  // Insertion header points to the first block that has free tuple slots
+  // Once a txn arrives, it will start from the insertion header to find the first
+  // idle (no other txn is trying to get tuple slots in that block) and non-full block.
+  // If no such block is found, the txn will create a new block.
+  // Before the txn writes to the block, it will set block status to busy.
+  // The first bit of block insert_head_ is used to indicate if the block is busy
+  // If the first bit is 1, it indicates one txn is writing to the block.
+
   TupleSlot result;
+  auto block = insertion_head_;
   while (true) {
-    RawBlock *block = insertion_head_.load();
-    if (block != nullptr && accessor_.Allocate(block, &result)) break;
-    NewBlock(block);
+    // No free block left
+    if (block == blocks_.end()) {
+      RawBlock *new_block = NewBlock();
+      TERRIER_ASSERT(accessor_.SetBlockBusyStatus(new_block), "Status of new block should not be busy");
+      // No need to flip the busy status bit
+      accessor_.Allocate(new_block, &result);
+      // take latch
+      common::SpinLatch::ScopedSpinLatch guard(&blocks_latch_);
+      // insert block
+      blocks_.push_back(new_block);
+      block = --blocks_.end();
+      break;
+    }
+
+    if (accessor_.SetBlockBusyStatus(*block)) {
+      // No one is inserting into this block
+      if (accessor_.Allocate(*block, &result)) {
+        // The block is not full, succeed
+        break;
+      }
+      // Fail to insert into the block, flip back the status bit
+      accessor_.ClearBlockBusyStatus(*block);
+      // if the full block is the insertion_header, move the insertion_header
+      // Next insert txn will search from the new insertion_header
+      CheckMoveHead(block);
+      // The block is full, try next block
+      ++block;
+    } else {  // The block is inserting by other txn
+      // Try next block
+      ++block;
+    }
   }
+
+  // Do not need to wait unit finish inserting,
+  // can flip back the status bit once the thread gets the allocated tuple slot
+  accessor_.ClearBlockBusyStatus(*block);
```

Comment:
I may have initially asked this on Slack, but is there any significant performance difference between releasing the lock before or after `InsertInto`?  There's potential cache-thrashing throughout the insert process (not just in the header, but with each attribute's bitmap and value as well), but it may not be significant in practice if the lock is held long enough to space the threads out.